File: sdks/sdkwork-craw-chat-sdk/sdkwork-craw-chat-sdk-typescript/composed/src/media-module.ts
import type {
  AttachMediaRequest,
  AttachTextMediaOptions,
  CrawChatMediaUploadOptions,
  CrawChatUploadBody,
  CrawChatUploadFetchLike,
  CompleteUploadRequest,
  CreateUploadRequest,
  MediaAsset,
  MediaDownloadUrlResponse,
  MediaUploadMutationResponse,
  MediaUploadSession,
  PostMessageResult,
  QueryParams,
} from './types.js';
import type { CrawChatSdkContext } from './sdk-context.js';

function resolveUploadFetch(fetchOverride?: CrawChatUploadFetchLike): CrawChatUploadFetchLike {
  const fetchImpl = fetchOverride ?? globalThis.fetch;
  if (typeof fetchImpl !== 'function') {
    throw new Error(
      'CrawChat media upload requires a fetch implementation. Pass options.fetch or use an environment with global fetch.',
    );
  }
  return fetchImpl as CrawChatUploadFetchLike;
}

function requireUploadSession(response: MediaUploadMutationResponse): MediaUploadSession {
  if (response.upload) {
    return response.upload;
  }
  throw new Error(
    `Media asset ${response.mediaAssetId} did not include a presigned upload session.`,
  );
}

function buildCompleteUploadRequest(
  upload: MediaUploadSession,
  checksum?: string,
): CompleteUploadRequest {
  return {
    bucket: upload.bucket,
    objectKey: upload.objectKey,
    storageProvider: upload.storageProvider,
    url: upload.url,
    checksum,
  };
}

async function assertUploadSucceeded(response: Awaited<ReturnType<CrawChatUploadFetchLike>>) {
  if (response.ok) {
    return;
  }

  let detail = '';
  try {
    detail = (await response.text()).trim();
  } catch {
    detail = '';
  }

  const suffix = detail ? `: ${detail}` : '';
  throw new Error(`CrawChat media upload failed with status ${response.status}${suffix}`);
}

export class CrawChatMediaModule {
  constructor(private readonly context: CrawChatSdkContext) {}

  createUpload(body: CreateUploadRequest): Promise<MediaUploadMutationResponse> {
    return this.context.backendClient.media.createMediaUpload(body);
  }

  completeUpload(
    mediaAssetId: string | number,
    body: CompleteUploadRequest,
  ): Promise<MediaUploadMutationResponse> {
    return this.context.backendClient.media.completeMediaUpload(mediaAssetId, body);
  }

  async uploadContent(
    upload: MediaUploadSession,
    body: CrawChatUploadBody,
    options: CrawChatMediaUploadOptions = {},
  ): Promise<void> {
    const fetchImpl = resolveUploadFetch(options.fetch);
    const response = await fetchImpl(upload.url, {
      method: upload.method,
      headers: upload.headers,
      body,
    });
    await assertUploadSucceeded(response);
  }

  async upload(
    request: CreateUploadRequest,
    body: CrawChatUploadBody,
    options: CrawChatMediaUploadOptions = {},
  ): Promise<MediaUploadMutationResponse> {
    const created = await this.createUpload(request);
    const upload = requireUploadSession(created);
    await this.uploadContent(upload, body, options);
    return this.completeUpload(
      created.mediaAssetId,
      buildCompleteUploadRequest(upload, options.checksum),
    );
  }

  getDownloadUrl(
    mediaAssetId: string | number,
    params?: QueryParams,
  ): Promise<MediaDownloadUrlResponse> {
    return this.context.backendClient.media.getMediaDownloadUrl(mediaAssetId, params);
  }

  get(mediaAssetId: string | number): Promise<MediaAsset> {
    return this.context.backendClient.media.getMediaAsset(mediaAssetId);
  }

  attach(
    mediaAssetId: string | number,
    body: AttachMediaRequest,
  ): Promise<PostMessageResult> {
    return this.context.backendClient.media.attachMediaAsset(mediaAssetId, body);
  }

  attachText(
    mediaAssetId: string | number,
    options: AttachTextMediaOptions,
  ): Promise<PostMessageResult> {
    return this.attach(mediaAssetId, {
      conversationId: options.conversationId,
      clientMsgId: options.clientMsgId,
      summary: options.summary,
      text: options.text,
      renderHints: options.renderHints,
    });
  }
}

File: sdks/sdkwork-craw-chat-sdk/sdkwork-craw-chat-sdk-rust/composed/src/media_module.rs
use crate::{CrawChatSdkContext, QueryParams};
use sdkwork_craw_chat_backend_sdk::{
  AttachMediaRequest,
  CompleteUploadRequest,
  CreateUploadRequest,
  MediaAsset,
  MediaDownloadUrlResponse,
  PostMessageResult,
  SdkworkError,
};

#[derive(Clone)]
pub struct CrawChatMediaModule {
  context: CrawChatSdkContext,
}

impl CrawChatMediaModule {
  pub(crate) fn new(context: CrawChatSdkContext) -> Self {
    Self { context }
  }

  pub async fn create_upload(
    &self,
    body: CreateUploadRequest,
  ) -> Result<MediaAsset, SdkworkError> {
    self.context.backend_client().media().create_media_upload(&body).await
  }

  pub async fn complete_upload(
    &self,
    media_asset_id: impl AsRef<str>,
    body: CompleteUploadRequest,
  ) -> Result<MediaAsset, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .backend_client()
      .media()
      .complete_media_upload(&media_asset_id, &body)
      .await
  }

  pub async fn download_url(
    &self,
    media_asset_id: impl AsRef<str>,
    params: Option<&QueryParams>,
  ) -> Result<MediaDownloadUrlResponse, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .backend_client()
      .media()
      .get_media_download_url(&media_asset_id, params)
      .await
  }

  pub async fn get(
    &self,
    media_asset_id: impl AsRef<str>,
  ) -> Result<MediaAsset, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self.context.backend_client().media().get_media_asset(&media_asset_id).await
  }

  pub async fn attach(
    &self,
    media_asset_id: impl AsRef<str>,
    body: AttachMediaRequest,
  ) -> Result<PostMessageResult, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .backend_client()
      .media()
      .attach_media_asset(&media_asset_id, &body)
      .await
  }
}
